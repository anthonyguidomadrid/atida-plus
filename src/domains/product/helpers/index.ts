export * from './calculate-discount'
export * from './calculate-price-per-unit'
export * from './get-sale-price'
