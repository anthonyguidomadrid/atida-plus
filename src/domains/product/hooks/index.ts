export * from './useFormatPrice'
export * from './useFormatAmount'
export * from './useFormatWeight'
