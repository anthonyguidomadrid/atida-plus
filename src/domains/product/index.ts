export * from './helpers'
export * from './hooks'
export * from './slices'
export * from './sagas'
export * from './selectors'
export * from './types'
export * from './normalizers'
