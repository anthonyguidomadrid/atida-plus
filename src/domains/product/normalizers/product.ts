import { removeUndefinedPropertiesFromObject, sanitiseUrl } from '~helpers'
import {
  calculateDiscount,
  calculatePricePerUnit,
  getSalePrice
} from '../helpers'
import {
  AlgoliaProduct,
  ChannelPrice,
  ElasticSearchProduct,
  InfoLabelEnum,
  Product,
  ProductImageDerivative,
  ProductLabelWrapper,
  RawProductAttributes
} from '../types'
import { HIDE_LABEL } from '~config/constants/promotions'
import { getLocaleKey, getEnabledLocales } from '~domains/translated-routes'
import { SessionChannelType } from '~domains/basket/types'

const formatSlugAsUrl = (slug?: string) => (slug ? `/${slug}` : undefined)

const getUrlSlugForLocale = (
  locale?: string,
  attributes?: Partial<RawProductAttributes>
) => {
  const localeKey = getLocaleKey(locale)

  if (localeKey && attributes?.hasOwnProperty('url_slug_' + localeKey)) {
    // @ts-ignore
    const slugAttribute = attributes[`url_slug_${localeKey}`] ?? ''
    return formatSlugAsUrl(slugAttribute)
  }

  return undefined
}

const returnAlternativeUrls = (attributes?: Partial<RawProductAttributes>) => {
  const alternativeUrlObject = {}

  getEnabledLocales().map(locale => {
    const localeKey = getLocaleKey(locale)

    if (localeKey && attributes?.hasOwnProperty('url_slug_' + localeKey)) {
      // @ts-ignore
      alternativeUrlObject[`${locale}`] =
        // @ts-ignore
        attributes[`url_slug_${localeKey}`] ?? ''
    }
  })

  return alternativeUrlObject
}

const makeProductLabels = (
  campaign: string,
  promotions: string[],
  pricingTag?: { code: string; label: string },
  price?: { sale?: number; rrp?: number },
  isPromotionalItem?: boolean,
  isFullyDiscountedItem?: boolean,
  isSample?: boolean
): ProductLabelWrapper[] | undefined => {
  const labels = [
    campaign && !isPromotionalItem
      ? {
          type: InfoLabelEnum.CampaignPromotion,
          label: `${InfoLabelEnum.CampaignPromotion}.${campaign}`
        }
      : null,
    ...(isPromotionalItem
      ? [
          {
            type: InfoLabelEnum.CampaignPromotion,
            label: isFullyDiscountedItem
              ? isSample
                ? 'campaign-promo.sample'
                : 'campaign-promo.gift'
              : 'third-rank.discounted-product'
          }
        ]
      : formatPromotionLabel(promotions)),
    pricingTag
      ? {
          type:
            pricingTag.code === 'outlet'
              ? InfoLabelEnum.PricingTagOutlet
              : InfoLabelEnum.PricingTagMini,
          label: pricingTag.label
        }
      : null,
    price?.sale && price?.rrp && price.sale < price.rrp
      ? {
          type: InfoLabelEnum.Discount,
          label: `-${calculateDiscount(price.sale, price.rrp)}%`
        }
      : null
  ].filter((label): label is ProductLabelWrapper => !!label)

  if (labels.length === 0) {
    return undefined
  }

  return labels
}

// filters promotions with hidelabel content first,
// then takes the last one of the list (not the first one anymore)
// and formats it finally in the desired structure
export const formatPromotionLabel = (promotions: string[]) => {
  return promotions
    .filter((p: string) => !p.includes(HIDE_LABEL))
    .slice(-1)
    .map(promotion => ({
      type: InfoLabelEnum.Promotion,
      label: promotion
    }))
}

const getProductImageUrl = (
  url: string | undefined,
  derivative: ProductImageDerivative,
  isPromotionalItem?: boolean
) => {
  if (url) return sanitiseUrl(url)
  return `/images/no-image-${
    isPromotionalItem ? 'Gift' : 'Product'
  }%20${derivative}.png`
}

export const normalizeElasticSearchProduct = (
  locale: string,
  item?: ElasticSearchProduct,
  sessionChannel?: SessionChannelType,
  mockedProductPriceChannels?: ChannelPrice[],
  isPromotionalItem?: boolean,
  isFullyDiscountedItem?: boolean
): Partial<Product> => {
  const salePrice = getSalePrice(
    { sku: item?._id, sale: item?._source?.price?.sale },
    mockedProductPriceChannels ?? item?._source?.channel_prices,
    sessionChannel
  )
  return removeUndefinedPropertiesFromObject({
    id: item?.id,
    name: item?._source?.name,
    enabled: item?._source?.enabled,
    sku: item?._id,
    pzn: item?._source?.attributes?.pzn,
    family: item?._source?.family,
    gtin: item?._source?.attributes?.ean,
    url: getUrlSlugForLocale(locale, item?._source?.attributes),
    alternativeUrls: removeUndefinedPropertiesFromObject(
      returnAlternativeUrls(item?._source?.attributes)
    ),
    quantityLeft: item?._source?.qty,
    categories: item?._source?.attributes?.categories,
    brand: item?._source?.attributes?.brand,
    format: item?._source?.attributes?.format,
    maxQuantity: item?._source?.attributes?.max_qty,
    productDatImage: item?._id
      ? item?._source?.attributes?.image_derivatives?.[0]?.dat_url
      : undefined,
    thumbnailImage: item?._id
      ? getProductImageUrl(
          item?._source?.attributes?.image_derivatives?.[0]
            ?.derivative_product_tile_thumbnail,
          ProductImageDerivative.Thumbnail,
          isPromotionalItem
        )
      : undefined,
    mediumImage: item?._id
      ? getProductImageUrl(
          item?._source?.attributes?.image_derivatives?.[0]
            ?.derivative_product_medium,
          ProductImageDerivative.Medium,
          isPromotionalItem
        )
      : undefined,
    largeImage: item?._id
      ? getProductImageUrl(
          item?._source?.attributes?.image_derivatives?.[0]
            ?.derivative_product_high,
          ProductImageDerivative.High,
          isPromotionalItem
        )
      : undefined,
    productTileImage: item?._id
      ? getProductImageUrl(
          item?._source?.attributes?.image_derivatives?.[0]
            ?.derivative_product_tile,
          ProductImageDerivative.ProductTile,
          isPromotionalItem
        )
      : undefined,
    productTileRetinaImage: item?._id
      ? getProductImageUrl(
          item?._source?.attributes?.image_derivatives?.[0]
            ?.derivative_product_tile_retina,
          ProductImageDerivative.ProductTileRetina,
          isPromotionalItem
        )
      : undefined,

    images: item?._id
      ? item?._source?.attributes?.image_derivatives?.map(image => {
          return {
            productDatImage: image?.dat_url,
            productHigh: getProductImageUrl(
              image.derivative_product_high,
              ProductImageDerivative.High,
              isPromotionalItem
            ),
            productTileThumbnail: getProductImageUrl(
              image.derivative_product_tile_thumbnail,
              ProductImageDerivative.Thumbnail,
              isPromotionalItem
            )
          }
        })
      : undefined,

    videos: item?._id
      ? item?._source?.attributes?.asset_video_list?.map(video => {
          return {
            video_id: video.video_id,
            label: video.label,
            type: video.type
          }
        })
      : undefined,

    description: item?._source?.attributes?.description_long,
    shortDescription: item?._source?.attributes?.description_short,
    hints: item?._source?.attributes?.notes,
    lmivStorage: item?._source?.attributes?.lmiv_storage,
    metaTitle: item?._source?.attributes?.meta_title,
    metaKeywords: item?._source?.attributes?.meta_keyword,
    metaDescription: item?._source?.attributes?.meta_description,
    unitVolume: item?._source?.attributes?.content_size
      ? {
          unit: item?._source?.attributes?.content_size?.unit,
          amount: item?._source?.attributes?.content_size?.amount,
          unitLabel: item?._source?.attributes?.content_size?.unit_label
        }
      : undefined,
    contentSizeFactor: item?._source?.attributes?.content_size_factor,
    bundleSizeFactor: item?._source?.attributes?.bundle_size_factor,
    usageNotes: item?._source?.attributes?.usage_notes,
    ingredients: item?._source?.attributes?.ingredient_list,
    assetCertificateList: item?._source?.attributes?.asset_certificate_list,
    price: item?._source?.price
      ? {
          currency: item?._source?.price?.currency,
          value: salePrice
        }
      : undefined,
    pricePerUnit:
      item?._source?.price && item?._source?.attributes?.content_size
        ? removeUndefinedPropertiesFromObject(
            calculatePricePerUnit({
              locale,
              price: { ...item._source.price, sale: salePrice },
              contentSizeFactor: item?._source?.attributes?.content_size_factor,
              contentSize: item._source.attributes.content_size
            })
          )
        : undefined,
    rrp: item?._source?.price
      ? {
          value: item?._source?.price?.rrp,
          currency: item?._source?.price?.currency
        }
      : undefined,
    availability: item?._source?.availability_state,
    labels: makeProductLabels(
      item?._source?.campaign ?? '',
      item?._source?.promos ?? [],
      item?._source?.attributes?.tag,
      {
        sale: salePrice,
        rrp: item?._source?.price?.rrp
      },
      isPromotionalItem,
      isFullyDiscountedItem,
      item?._source?.attributes?.not_for_sale_type?.code === 'sample'
    ),
    pharmaceuticalAdvice: item?._source?.attributes?.pharmaceutical_advice,
    rating: item?._source?.review
      ? {
          averageRating: item._source.review.product_score,
          numberOfReviews: item._source.review.total_reviews
        }
      : undefined,
    notForSaleType: item?._source?.attributes?.not_for_sale_type?.code,
    isOnDemand: item?._source?.attributes?.is_on_demand,
    manufacturerName: item?._source?.attributes?.manufacturer_name,
    netWeight: item?._source?.attributes?.net_weight
      ? {
          unit: item?._source?.attributes?.net_weight?.unit,
          amount: item?._source?.attributes?.net_weight?.amount,
          unitLabel: item?._source?.attributes?.net_weight?.unit_label
        }
      : undefined,
    isBiocide: item?._source?.attributes?.is_biocide,
    biocideBauaNumber: item?._source?.attributes?.biocide_baua_number,
    isReimport: item?._source?.attributes?.is_reimport
      ? {
          code: item?._source?.attributes?.is_reimport?.code,
          label: item?._source?.attributes?.is_reimport?.label
        }
      : undefined,
    organicControlNumber:
      item?._source?.attributes?.organic_control_body_number,
    hasOrganicSealLogo: item?._source?.attributes?.has_organic_seal_logo
      ? {
          code: item?._source?.attributes?.has_organic_seal_logo?.code,
          label: item?._source?.attributes?.has_organic_seal_logo?.label
        }
      : undefined,
    ethanolContent: item?._source?.attributes?.ethanol_content,
    isPharmacyExclusive: item?._source?.attributes?.is_pharmacy_exclusive,
    productFeaturesListMultiselect: item?._source?.attributes
      ?.product_features_list_multiselect
      ? [...item?._source?.attributes?.product_features_list_multiselect]
      : undefined,
    dietetic: item?._source?.attributes?.dietetic
      ? {
          code: item?._source?.attributes?.dietetic?.code,
          label: item?._source?.attributes?.dietetic?.label
        }
      : undefined
  })
}

export const normalizeAlgoliaProduct = (
  locale?: string,
  item?: AlgoliaProduct,
  sessionChannel?: SessionChannelType,
  mockedProductPriceChannels?: ChannelPrice[]
): Partial<Product> => {
  const salePrice = getSalePrice(
    { sku: item?.sku, sale: item?.price?.sale },
    mockedProductPriceChannels ?? item?.channel_prices,
    sessionChannel
  )
  return removeUndefinedPropertiesFromObject({
    name: item?.name,
    sku: item?.sku,
    url: getUrlSlugForLocale(locale, item?.attributes),
    brand: item?.attributes?.brand,
    format: item?.attributes?.format,
    productDatImage: item?.sku
      ? item?.attributes?.image_derivatives?.[0]?.dat_url
      : undefined,
    thumbnailImage: item?.sku
      ? getProductImageUrl(
          item?.attributes?.image_derivatives?.[0]
            ?.derivative_product_tile_thumbnail,
          ProductImageDerivative.Thumbnail
        )
      : undefined,
    mediumImage: item?.sku
      ? getProductImageUrl(
          item?.attributes?.image_derivatives?.[0]?.derivative_product_medium,
          ProductImageDerivative.Medium
        )
      : undefined,
    largeImage: item?.sku
      ? getProductImageUrl(
          item?.attributes?.image_derivatives?.[0]?.derivative_product_high,
          ProductImageDerivative.High
        )
      : undefined,
    productTileImage: item?.sku
      ? getProductImageUrl(
          item?.attributes?.image_derivatives?.[0]?.derivative_product_tile,
          ProductImageDerivative.ProductTile
        )
      : undefined,
    productTileRetinaImage: item?.sku
      ? getProductImageUrl(
          item?.attributes?.image_derivatives?.[0]
            ?.derivative_product_tile_retina,
          ProductImageDerivative.ProductTileRetina
        )
      : undefined,
    description: item?.attributes?.description_long,
    shortDescription: item?.attributes?.description_short,
    categories: item?.attributes?.categories,
    tag: item?.attributes?.tag,
    metaTitle: item?.attributes?.meta_title,
    metaKeywords: item?.attributes?.meta_keyword,
    metaDescription: item?.attributes?.meta_description,
    unitVolume: item?.attributes?.content_size
      ? {
          unit: item?.attributes?.content_size?.unit,
          amount: item?.attributes?.content_size?.amount,
          unitLabel: item?.attributes?.content_size?.unit_label
        }
      : undefined,
    contentSizeFactor: item?.attributes?.content_size_factor,
    price: item?.price
      ? { currency: item?.price?.currency, value: salePrice }
      : undefined,
    pricePerUnit:
      item?.price && item?.attributes?.content_size
        ? removeUndefinedPropertiesFromObject(
            calculatePricePerUnit({
              locale,
              price: { ...item.price, sale: salePrice },
              contentSizeFactor: item?.attributes?.content_size_factor,
              contentSize: item.attributes.content_size
            })
          )
        : undefined,
    rrp: item?.sku
      ? {
          value: item?.price?.rrp,
          currency: item?.price?.currency
        }
      : undefined,
    availability: item?.availability?.state,
    labels: makeProductLabels(
      item?.campaign ?? '',
      item?.promos ?? [],
      item?.attributes?.tag,
      {
        sale: salePrice,
        rrp: item?.price?.rrp
      }
    ),
    rating: item?.review
      ? {
          averageRating: item.review.product_score,
          numberOfReviews: item.review.total_reviews
        }
      : undefined
  })
}

export const normalizeElasticSearchMinimalProduct = (
  locale?: string,
  item?: ElasticSearchProduct,
  sessionChannel?: SessionChannelType,
  mockedProductPriceChannels?: ChannelPrice[],
  isPromotionalItem?: boolean,
  isFullyDiscountedItem?: boolean
): Partial<Product> => {
  const salePrice = getSalePrice(
    { sku: item?._id, sale: item?._source?.price?.sale },
    mockedProductPriceChannels ?? item?._source?.channel_prices,
    sessionChannel
  )
  return removeUndefinedPropertiesFromObject({
    id: item?.id,
    name: item?._source?.name,
    enabled: item?._source?.enabled,
    sku: item?._id,
    url: getUrlSlugForLocale(locale, item?._source?.attributes),
    alternativeUrls: removeUndefinedPropertiesFromObject(
      returnAlternativeUrls(item?._source?.attributes)
    ),
    format: item?._source?.attributes?.format,
    maxQuantity: item?._source?.attributes?.max_qty,
    productDatImage: item?._id
      ? item?._source?.attributes?.image_derivatives?.[0]?.dat_url
      : undefined,
    thumbnailImage: item?._id
      ? getProductImageUrl(
          item?._source?.attributes?.image_derivatives?.[0]
            ?.derivative_product_tile_thumbnail,
          ProductImageDerivative.Thumbnail,
          isPromotionalItem
        )
      : undefined,
    mediumImage: item?._id
      ? getProductImageUrl(
          item?._source?.attributes?.image_derivatives?.[0]
            ?.derivative_product_medium,
          ProductImageDerivative.Medium,
          isPromotionalItem
        )
      : undefined,
    largeImage: item?._id
      ? getProductImageUrl(
          item?._source?.attributes?.image_derivatives?.[0]
            ?.derivative_product_high,
          ProductImageDerivative.High,
          isPromotionalItem
        )
      : undefined,
    productTileImage: item?._id
      ? getProductImageUrl(
          item?._source?.attributes?.image_derivatives?.[0]
            ?.derivative_product_tile,
          ProductImageDerivative.ProductTile,
          isPromotionalItem
        )
      : undefined,
    productTileRetinaImage: item?._id
      ? getProductImageUrl(
          item?._source?.attributes?.image_derivatives?.[0]
            ?.derivative_product_tile_retina,
          ProductImageDerivative.ProductTileRetina,
          isPromotionalItem
        )
      : undefined,
    metaTitle: item?._source?.attributes?.meta_title,
    metaKeywords: item?._source?.attributes?.meta_keyword,
    metaDescription: item?._source?.attributes?.meta_description,
    unitVolume: item?._source?.attributes?.content_size
      ? {
          unit: item?._source?.attributes?.content_size?.unit,
          amount: item?._source?.attributes?.content_size?.amount,
          unitLabel: item?._source?.attributes?.content_size?.unit_label
        }
      : undefined,
    price: item?._source?.price
      ? {
          currency: item?._source?.price?.currency,
          value: salePrice
        }
      : undefined,
    pricePerUnit:
      item?._source?.price && item?._source?.attributes?.content_size
        ? removeUndefinedPropertiesFromObject(
            calculatePricePerUnit({
              locale,
              price: { ...item._source.price, sale: salePrice },
              contentSizeFactor: item?._source?.attributes?.content_size_factor,
              contentSize: item._source.attributes.content_size
            })
          )
        : undefined,
    rrp: item?._source?.price
      ? {
          value: item?._source?.price?.rrp,
          currency: item?._source?.price?.currency
        }
      : undefined,
    availability: item?._source?.availability_state,
    labels: makeProductLabels(
      item?._source?.campaign ?? '',
      item?._source?.promos ?? [],
      item?._source?.attributes?.tag,
      {
        sale: salePrice,
        rrp: item?._source?.price?.rrp
      },
      isPromotionalItem,
      isFullyDiscountedItem,
      item?._source?.attributes?.not_for_sale_type?.code === 'sample'
    ),
    rating: item?._source?.review
      ? {
          averageRating: item._source.review.product_score,
          numberOfReviews: item._source.review.total_reviews
        }
      : undefined
  })
}
