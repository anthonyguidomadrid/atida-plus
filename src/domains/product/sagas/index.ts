import type { SagaIterator } from 'redux-saga'
import { all, fork } from 'redux-saga/effects'
import { getProductSaga } from './get-product'

export function* productSaga(): SagaIterator {
  yield all([getProductSaga].map(fork))
}
