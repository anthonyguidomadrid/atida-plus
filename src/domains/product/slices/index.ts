import { combineReducers } from '@reduxjs/toolkit'

import { contentSlice } from './content'
import { viewToggleSlice } from './view-toggle'

export const productReducer = combineReducers({
  content: contentSlice.reducer,
  viewToggle: viewToggleSlice.reducer
})

export const {
  trigger: contentTrigger,
  request: contentRequest,
  success: contentSuccess,
  failure: contentFailure,
  fulfill: contentFulfill,
  setShowLabels: setShowLabels,
  setClearData: setClearData
} = contentSlice.actions

export const { toggleView } = viewToggleSlice.actions
