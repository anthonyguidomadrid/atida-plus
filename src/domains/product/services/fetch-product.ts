import {
  createClient,
  transformLocaleToUppercase,
  getCountryTagFromLocale
} from '~helpers'
import getConfig from 'next/config'
import {
  createElasticSearchClient,
  getElasticSearchProductIndex
} from '~domains/elasticsearch/helpers/client'
import { normalizeElasticSearchProduct } from '../normalizers'
import {
  Product,
  ElasticSearchProduct,
  GetElasticSearchProductPayload,
  ChannelPrice
} from '../types'
import { cache } from '~helpers/server-only/redisClient'
import { ProductBrandQuery, ProductBrand } from '~generated-graphql'
import { getAvailableLocales } from '~domains/translated-routes'
import { AxiosResponse } from 'axios'
import { loadFeatureFlag } from '~helpers/server-only/featureFlagClient'
import { FeatureFlag } from '~config/constants/feature-flags'

if (typeof window !== 'undefined')
  throw new Error('This file should not be called client-side')

const { serverRuntimeConfig } = getConfig()
const client = createElasticSearchClient()

const normalizeProductBrandResponse = (
  response: AxiosResponse<{
    data: ProductBrandQuery
  }>
): string =>
  response?.data?.data?.brandCollection?.items?.[0]?.linkedFrom?.pageCollection
    ?.items?.[0]?.slug ?? ''

const fetchProductBrand = async (locale: string, brand: string) => {
  const contentfulClient = createClient({
    options: {
      baseURL: `${serverRuntimeConfig.cmsGraphQLHost}/spaces/${serverRuntimeConfig.cmsSpaceId}/environments/${serverRuntimeConfig.cmsEnvironmentId}`,
      headers: {
        Authorization: `Bearer ${serverRuntimeConfig.cmsToken}`
      },
      adapter: cache?.adapter
    },
    locale,
    interceptorOptions: {
      mode: 'contentful-graphql'
    }
  })
  const contentfulResponse = await contentfulClient.post<{
    data: ProductBrandQuery
  }>('', {
    query: ProductBrand.loc?.source.body,
    variables: {
      id: brand,
      allLocales: getAvailableLocales().map(transformLocaleToUppercase),
      locale: transformLocaleToUppercase(locale),
      tags: getCountryTagFromLocale(locale)
    }
  })

  return normalizeProductBrandResponse(contentfulResponse)
}

export const fetchProduct = async (
  locale: string,
  data: GetElasticSearchProductPayload
): Promise<Partial<Product>> => {
  const fetchSingleBrand = await loadFeatureFlag(
    locale,
    FeatureFlag.PRODUCT_PDP_FETCH_SINGLE_BRAND
  )
  // TODO: Remove when the price channels on products are ready to be used
  const mockedProductPriceChannels = (await loadFeatureFlag(
    locale,
    FeatureFlag.MUTLICHANNEL_PRICES_MOCKED_PRODUCT_PRICE_CHANNEL
  )) as { channel_prices: ChannelPrice[] }

  const response = await client.get<ElasticSearchProduct>({
    index: getElasticSearchProductIndex(locale),
    id: data?.sku ?? ''
  })
  const product = normalizeElasticSearchProduct(
    locale,
    response.body,
    data?.sessionChannel,
    mockedProductPriceChannels?.channel_prices
  )

  if (fetchSingleBrand && product?.brand?.code) {
    const brandUrl = await fetchProductBrand(locale, product?.brand?.code)
    return { ...product, brandUrl }
  }

  return product
}
