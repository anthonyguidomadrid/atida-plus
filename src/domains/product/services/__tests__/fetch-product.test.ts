/**
 * @jest-environment node
 */
// @ts-ignore
import { get } from '@elastic/elasticsearch'
import { elasticsearchProduct } from '../../__mocks__/product'
import { fetchProduct } from '../fetch-product'

describe(fetchProduct, () => {
  beforeEach(() => {
    get.mockResolvedValue(elasticsearchProduct)
  })

  it('passes the correct index and sku to elasticsearch request', async () => {
    await fetchProduct('en-gb', {
      sku: '200000004',
      sessionChannel: { channel: 'some-channel', sku: 'some-sku' }
    })
    expect(get).toHaveBeenCalledWith({
      id: '200000004',
      index: 'index-en-gb'
    })
  })

  it('returns the normalized product response', async () => {
    const response = await fetchProduct('en-gb', {
      sku: '200000004',
      sessionChannel: { channel: 'some-channel', sku: 'some-sku' }
    })
    expect(response).toMatchSnapshot()
  })
})
