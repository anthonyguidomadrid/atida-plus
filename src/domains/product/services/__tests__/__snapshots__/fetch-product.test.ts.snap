// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchProduct returns the normalized product response 1`] = `
Object {
  "alternativeUrls": Object {
    "es-es": "",
    "pt-pt": "forcapil-anti-caida-del-cabello-arkopharma-30-capsulas-2+1",
  },
  "availability": "AVAILABLE",
  "brand": Object {
    "code": "atida",
    "label": "Atida",
  },
  "categories": Object {
    "lvl0": Array [
      "main",
    ],
  },
  "contentSizeFactor": 1,
  "description": "<p><strong>Forcapil Anti-Caída del Cabello Arkopharma 30 Cápsulas 2+1</strong>
</p><p>¡Recupera la fuerza y vitalidad de tu cabello con Forcapil!
</p><p>Indicado para hombres y mujeres, este nutricosmético previene y combate la caída del pelo.
</p><p>Fórmula exclusivamente con ingredientes 100% de origen natural.
</p><p>Tiene una alta concentración en biotina, y su composición con aminoácidos, vitaminas, minerales y plantas favorece el crecimiento de la melena.
</p><p>Se recomienda tomar en momentos en los que sufrimos una mayor pérdida de cabello, causada por estrés, menopausia, tras el parto, cambios de estación...
</p><p><strong>Tratamiento para 3 meses: 2 + 1 de REGALO.</strong>
</p>",
  "enabled": true,
  "format": Object {
    "code": "shampooCode",
    "label": "Shampoo",
  },
  "gtin": "8428148462962",
  "images": Array [
    Object {
      "productDatImage": "https://sandbox-atida.bynder.com/transform/db2e62c0-2c32-4e50-bee7-1a7a421323f1/ean_img_8833772266123_0",
      "productHigh": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20High-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
      "productTileThumbnail": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Tile%20Thumbnail-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
    },
  ],
  "labels": Array [
    Object {
      "label": "Free sponge",
      "type": "promotion",
    },
    Object {
      "label": "Outlet",
      "type": "pricing-tag-outlet",
    },
    Object {
      "label": "-11%",
      "type": "discount",
    },
  ],
  "largeImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20High-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
  "manufacturerName": "LOreal Deutschland",
  "maxQuantity": 50,
  "mediumImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Medium-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
  "metaDescription": "Meta description No contiene gluten. Apto para personas celíacas Meta description",
  "metaKeywords": "Meta keywords NaturGreen Algarroba Meta keywords",
  "metaTitle": "Meta title NaturGreen Algarroba a la Taza BIO 225gr Meta title",
  "name": "Forcapil Anti-Caída del Cabello Arkopharma 30 Cápsulas 2+1",
  "netWeight": Object {
    "amount": 100,
    "unit": "Grams",
    "unitLabel": "g",
  },
  "price": Object {
    "currency": "EUR",
    "value": 8888,
  },
  "pricePerUnit": Object {
    "currency": "EUR",
    "unit": "100 ml",
    "value": 8888,
  },
  "productDatImage": "https://sandbox-atida.bynder.com/transform/db2e62c0-2c32-4e50-bee7-1a7a421323f1/ean_img_8833772266123_0",
  "productTileImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Tile-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
  "productTileRetinaImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Tile%20Retina-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
  "pzn": "1234567",
  "rating": Object {
    "averageRating": 3.5,
    "numberOfReviews": 100,
  },
  "rrp": Object {
    "currency": "EUR",
    "value": 9999,
  },
  "shortDescription": "Preparado vegetal con harina de algarroba ecológica, ideal para desayunos y meriendas",
  "sku": "100000000",
  "thumbnailImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Tile%20Thumbnail-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
  "unitVolume": Object {
    "amount": 100,
    "unit": "MILLILITER",
    "unitLabel": "ml",
  },
  "url": "/forcapil-anti-caida-del-cabello-arkopharma-30-capsulas-2+1",
}
`;
